File: app/src/main/java/io/oldering/tvfoot/red/app/domain/matches/displayable/MatchesItemDisplayableDiffUtilCallback.java
package io.oldering.tvfoot.red.app.domain.matches.displayable;

import android.support.v7.util.DiffUtil;
import java.util.List;

public class MatchesItemDisplayableDiffUtilCallback extends DiffUtil.Callback {
  private List<MatchesItemDisplayable> oldItems;
  private List<MatchesItemDisplayable> newItems;

  public MatchesItemDisplayableDiffUtilCallback() {
  }

  @Override public int getOldListSize() {
    return oldItems.size();
  }

  @Override public int getNewListSize() {
    return newItems.size();
  }

  @Override public boolean areItemsTheSame(int oldItemPosition, int newItemPosition) {
    MatchesItemDisplayable oldItem = oldItems.get(oldItemPosition);
    MatchesItemDisplayable newItem = newItems.get(newItemPosition);

    return oldItem.isSameAs(newItem);
  }

  @Override public boolean areContentsTheSame(int oldItemPosition, int newItemPosition) {
    MatchesItemDisplayable oldItem = oldItems.get(oldItemPosition);
    MatchesItemDisplayable newItem = newItems.get(newItemPosition);

    return oldItem.equals(newItem);
  }

  public void bindItems(List<MatchesItemDisplayable> oldItems,
      List<MatchesItemDisplayable> newItems) {
    this.oldItems = oldItems;
    this.newItems = newItems;
  }
}

File: app/src/main/java/com/benoitquenaudon/tvfoot/red/app/domain/matches/displayable/MatchesItemDisplayableDiffUtilCallback.kt
package com.benoitquenaudon.tvfoot.red.app.domain.matches.displayable

import androidx.recyclerview.widget.DiffUtil

class MatchesItemDisplayableDiffUtilCallback<out T : MatchesItemDisplayable>(
    private val oldItems: List<T>,
    private val newItems: List<T>
) : DiffUtil.Callback() {

  override fun getOldListSize(): Int {
    return oldItems.size
  }

  override fun getNewListSize(): Int {
    return newItems.size
  }

  override fun areItemsTheSame(oldItemPosition: Int, newItemPosition: Int) =
      oldItems[oldItemPosition].isSameAs(newItems[newItemPosition])

  override fun areContentsTheSame(oldItemPosition: Int, newItemPosition: Int) =
      oldItems[oldItemPosition] == newItems[newItemPosition]
}
